File: ios/Runner/AudioEngine/MIDIEngine.swift
import CoreMIDI
import AVFoundation

/// CoreMIDI engine for GrooveLab.
/// Enables MIDI input/output for integration with MainStage, Logic Pro, Ableton, etc.
/// Supports MIDI Clock send/receive, Note input, CC mapping, and Network MIDI.
final class MIDIEngine {
    private var midiClient: MIDIClientRef = 0
    private var inputPort: MIDIPortRef = 0
    private var outputPort: MIDIPortRef = 0
    private var virtualSource: MIDIEndpointRef = 0
    private var virtualDestination: MIDIEndpointRef = 0
    private var connectedSources: Set<MIDIEndpointRef> = []

    // MIDI Clock
    private var clockTimer: DispatchSourceTimer?
    private var isSendingClock = false
    private var clockBpm: Double = 120.0

    // Callbacks
    private let onEvent: ([String: Any]) -> Void

    // State
    private var isInitialized = false

    init(onEvent: @escaping ([String: Any]) -> Void) {
        self.onEvent = onEvent
    }

    // MARK: - Initialization

    func initialize() -> Bool {
        guard !isInitialized else { return true }

        var status: OSStatus

        // Create MIDI client
        status = MIDIClientCreateWithBlock("GrooveLab" as CFString, &midiClient) { [weak self] notification in
            self?.handleMIDINotification(notification)
        }
        guard status == noErr else {
            print("MIDIEngine: Failed to create client: \(status)")
            return false
        }

        // Create input port for receiving MIDI
        status = MIDIInputPortCreateWithProtocol(
            midiClient,
            "GrooveLab Input" as CFString,
            ._1_0,
            &inputPort
        ) { [weak self] eventList, srcConnRefCon in
            self?.handleMIDIEventList(eventList)
        }
        guard status == noErr else {
            print("MIDIEngine: Failed to create input port: \(status)")
            return false
        }

        // Create output port for sending MIDI
        status = MIDIOutputPortCreate(midiClient, "GrooveLab Output" as CFString, &outputPort)
        guard status == noErr else {
            print("MIDIEngine: Failed to create output port: \(status)")
            return false
        }

        // Create virtual source (so other apps can receive from GrooveLab)
        status = MIDISourceCreateWithProtocol(
            midiClient,
            "GrooveLab" as CFString,
            ._1_0,
            &virtualSource
        )
        if status != noErr {
            print("MIDIEngine: Warning - Failed to create virtual source: \(status)")
        }

        // Create virtual destination (so other apps can send to GrooveLab)
        status = MIDIDestinationCreateWithProtocol(
            midiClient,
            "GrooveLab" as CFString,
            ._1_0,
            &virtualDestination
        ) { [weak self] eventList, srcConnRefCon in
            self?.handleMIDIEventList(eventList)
        }
        if status != noErr {
            print("MIDIEngine: Warning - Failed to create virtual destination: \(status)")
        }

        // Enable Network MIDI session
        let networkSession = MIDINetworkSession.default()
        networkSession.isEnabled = true
        networkSession.connectionPolicy = .anyone

        // Connect to all existing sources
        connectToAllSources()

        isInitialized = true
        print("MIDIEngine: Initialized with \(getSourceCount()) sources, \(getDestinationCount()) destinations")
        return true
    }

    // MARK: - Device Discovery

    func getSourceCount() -> Int {
        return MIDIGetNumberOfSources()
    }

    func getDestinationCount() -> Int {
        return MIDIGetNumberOfDestinations()
    }

    func listDevices() -> [[String: Any]] {
        var devices: [[String: Any]] = []

        // List sources (input devices)
        for i in 0..<MIDIGetNumberOfSources() {
            let endpoint = MIDIGetSource(i)
            if let info = getEndpointInfo(endpoint) {
                var device = info
                device["direction"] = "input"
                device["index"] = i
                device["connected"] = connectedSources.contains(endpoint)
                devices.append(device)
            }
        }

        // List destinations (output devices)
        for i in 0..<MIDIGetNumberOfDestinations() {
            let endpoint = MIDIGetDestination(i)
            if let info = getEndpointInfo(endpoint) {
                var device = info
                device["direction"] = "output"
                device["index"] = i
                devices.append(device)
            }
        }

        return devices
    }

    private func getEndpointInfo(_ endpoint: MIDIEndpointRef) -> [String: Any]? {
        guard endpoint != 0 else { return nil }

        var name: Unmanaged<CFString>?
        MIDIObjectGetStringProperty(endpoint, kMIDIPropertyDisplayName, &name)
        let displayName = (name?.takeRetainedValue() as String?) ?? "Unknown"

        var manufacturer: Unmanaged<CFString>?
        MIDIObjectGetStringProperty(endpoint, kMIDIPropertyManufacturer, &manufacturer)
        let mfr = (manufacturer?.takeRetainedValue() as String?) ?? ""

        var uniqueID: Int32 = 0
        MIDIObjectGetIntegerProperty(endpoint, kMIDIPropertyUniqueID, &uniqueID)

        return [
            "name": displayName,
            "manufacturer": mfr,
            "uniqueId": Int(uniqueID),
            "endpointRef": Int(endpoint),
        ]
    }

    // MARK: - Connection Management

    func connectToAllSources() {
        for i in 0..<MIDIGetNumberOfSources() {
            let source = MIDIGetSource(i)
            connectToSource(source)
        }
    }

    func connectToSource(_ source: MIDIEndpointRef) {
        guard source != virtualDestination else { return } // Don't connect to ourselves
        let status = MIDIPortConnectSource(inputPort, source, nil)
        if status == noErr {
            connectedSources.insert(source)
        }
    }

    func disconnectFromSource(_ source: MIDIEndpointRef) {
        MIDIPortDisconnectSource(inputPort, source)
        connectedSources.remove(source)
    }

    // MARK: - MIDI Sending

    func sendNoteOn(note: UInt8, velocity: UInt8, channel: UInt8 = 0) {
        let status: UInt8 = 0x90 | (channel & 0x0F)
        sendMessage([status, note, velocity])
    }

    func sendNoteOff(note: UInt8, channel: UInt8 = 0) {
        let status: UInt8 = 0x80 | (channel & 0x0F)
        sendMessage([status, note, 0])
    }

    func sendCC(controller: UInt8, value: UInt8, channel: UInt8 = 0) {
        let status: UInt8 = 0xB0 | (channel & 0x0F)
        sendMessage([status, controller, value])
    }

    func sendProgramChange(program: UInt8, channel: UInt8 = 0) {
        let status: UInt8 = 0xC0 | (channel & 0x0F)
        sendMessage([status, program])
    }

    private func sendMessage(_ bytes: [UInt8]) {
        // Send to all destinations
        for i in 0..<MIDIGetNumberOfDestinations() {
            let dest = MIDIGetDestination(i)
            sendToEndpoint(dest, bytes: bytes)
        }

        // Also send via virtual source
        if virtualSource != 0 {
            sendViaSource(bytes)
        }
    }

    private func sendToEndpoint(_ dest: MIDIEndpointRef, bytes: [UInt8]) {
        var packet = MIDIPacket()
        packet.timeStamp = mach_absolute_time()
        packet.length = UInt16(bytes.count)
        withUnsafeMutablePointer(to: &packet.data) { ptr in
            let raw = UnsafeMutableRawPointer(ptr)
            for (i, byte) in bytes.enumerated() {
                raw.storeBytes(of: byte, toByteOffset: i, as: UInt8.self)
            }
        }

        var packetList = MIDIPacketList(numPackets: 1, packet: packet)
        MIDISend(outputPort, dest, &packetList)
    }

    private func sendViaSource(_ bytes: [UInt8]) {
        var packet = MIDIPacket()
        packet.timeStamp = mach_absolute_time()
        packet.length = UInt16(bytes.count)
        withUnsafeMutablePointer(to: &packet.data) { ptr in
            let raw = UnsafeMutableRawPointer(ptr)
            for (i, byte) in bytes.enumerated() {
                raw.storeBytes(of: byte, toByteOffset: i, as: UInt8.self)
            }
        }

        var packetList = MIDIPacketList(numPackets: 1, packet: packet)
        MIDIReceived(virtualSource, &packetList)
    }

    // MARK: - MIDI Clock

    func startClock(bpm: Double) {
        stopClock()
        clockBpm = bpm
        isSendingClock = true

        // Send MIDI Start (0xFA)
        sendMessage([0xFA])

        // MIDI Clock sends 24 pulses per quarter note
        let pulsesPerBeat: Double = 24.0
        let intervalNs = UInt64((60.0 / bpm / pulsesPerBeat) * 1_000_000_000)

        let timer = DispatchSource.makeTimerSource(flags: .strict, queue: .global(qos: .userInteractive))
        timer.schedule(deadline: .now(), repeating: .nanoseconds(Int(intervalNs)), leeway: .nanoseconds(0))
        timer.setEventHandler { [weak self] in
            guard let self = self, self.isSendingClock else { return }
            // MIDI Clock tick (0xF8)
            self.sendMessage([0xF8])
        }
        timer.resume()
        clockTimer = timer
    }

    func stopClock() {
        isSendingClock = false
        clockTimer?.cancel()
        clockTimer = nil
        // Send MIDI Stop (0xFC)
        sendMessage([0xFC])
    }

    func updateClockBpm(_ bpm: Double) {
        guard isSendingClock else {
            clockBpm = bpm
            return
        }
        // Restart clock with new BPM
        startClock(bpm: bpm)
    }

    // MARK: - MIDI Receive Handling

    private func handleMIDIEventList(_ eventList: UnsafePointer<MIDIEventList>) {
        let list = eventList.pointee
        // Use the modern Swift iteration for MIDIEventList
        eventList.unsafeSequence().forEach { event in
            let wordCount = Int(event.pointee.wordCount)
            guard wordCount > 0 else { return }
            // Read words from the packet
            withUnsafePointer(to: event.pointee.words) { ptr in
                ptr.withMemoryRebound(to: UInt32.self, capacity: wordCount) { wordsPtr in
                    for i in 0..<wordCount {
                        processMIDI1Word(wordsPtr[i])
                    }
                }
            }
        }
    }

    private func processMIDI1Word(_ word: UInt32) {
        // Universal MIDI Packet format (UMP) for MIDI 1.0
        let messageType = (word >> 28) & 0x0F
        guard messageType == 0x02 else { return } // MIDI 1.0 Channel Voice

        let status = UInt8((word >> 16) & 0xFF)
        let data1 = UInt8((word >> 8) & 0x7F)
        let data2 = UInt8(word & 0x7F)

        let messageKind = status & 0xF0
        let channel = status & 0x0F

        var eventDict: [String: Any] = [
            "type": "midi",
            "status": Int(status),
            "channel": Int(channel),
            "data1": Int(data1),
            "data2": Int(data2),
        ]

        switch messageKind {
        case 0x90: // Note On
            if data2 > 0 {
                eventDict["kind"] = "noteOn"
                eventDict["note"] = Int(data1)
                eventDict["velocity"] = Int(data2)
            } else {
                // Note On with velocity 0 = Note Off
                eventDict["kind"] = "noteOff"
                eventDict["note"] = Int(data1)
                eventDict["velocity"] = 0
            }
        case 0x80: // Note Off
            eventDict["kind"] = "noteOff"
            eventDict["note"] = Int(data1)
            eventDict["velocity"] = Int(data2)
        case 0xB0: // Control Change
            eventDict["kind"] = "cc"
            eventDict["controller"] = Int(data1)
            eventDict["value"] = Int(data2)
        case 0xC0: // Program Change
            eventDict["kind"] = "programChange"
            eventDict["program"] = Int(data1)
        case 0xE0: // Pitch Bend
            let bendValue = (Int(data2) << 7) | Int(data1)
            eventDict["kind"] = "pitchBend"
            eventDict["value"] = bendValue - 8192 // center at 0
        default:
            eventDict["kind"] = "other"
        }

        // Real-time messages (system)
        if status == 0xF8 { // Clock
            eventDict["kind"] = "clock"
        } else if status == 0xFA { // Start
            eventDict["kind"] = "start"
        } else if status == 0xFB { // Continue
            eventDict["kind"] = "continue"
        } else if status == 0xFC { // Stop
            eventDict["kind"] = "stop"
        }

        DispatchQueue.main.async { [weak self] in
            self?.onEvent(eventDict)
        }
    }

    // MARK: - MIDI Setup Change Notification

    private func handleMIDINotification(_ notification: UnsafePointer<MIDINotification>) {
        let messageID = notification.pointee.messageID

        switch messageID {
        case .msgSetupChanged, .msgObjectAdded, .msgObjectRemoved:
            // Reconnect to any new sources
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) { [weak self] in
                self?.connectToAllSources()
                self?.onEvent([
                    "type": "midiDeviceChange",
                    "devices": self?.listDevices() ?? [],
                ])
            }
        default:
            break
        }
    }

    // MARK: - Cleanup

    func dispose() {
        stopClock()

        for source in connectedSources {
            MIDIPortDisconnectSource(inputPort, source)
        }
        connectedSources.removeAll()

        if virtualSource != 0 { MIDIEndpointDispose(virtualSource) }
        if virtualDestination != 0 { MIDIEndpointDispose(virtualDestination) }
        if inputPort != 0 { MIDIPortDispose(inputPort) }
        if outputPort != 0 { MIDIPortDispose(outputPort) }
        if midiClient != 0 { MIDIClientDispose(midiClient) }

        isInitialized = false
        print("MIDIEngine: Disposed")
    }
}

File: android/app/src/main/kotlin/com/groovelab/groovelab/audio/MIDIEngine.kt
package com.groovelab.groovelab.audio

import android.content.Context
import android.media.midi.*
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.util.Log
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/**
 * Android MIDI engine for GrooveLab.
 * Supports MIDI input/output, device discovery, MIDI clock send/receive.
 * Compatible with USB MIDI, Bluetooth MIDI, and virtual MIDI.
 */
class MIDIEngine(
    private val context: Context,
    private val onEvent: (Map<String, Any>) -> Unit
) {
    companion object {
        private const val TAG = "MIDIEngine"
    }

    private var midiManager: MidiManager? = null
    private var openDevices: MutableList<MidiDevice> = mutableListOf()
    private var inputPorts: MutableList<MidiInputPort> = mutableListOf()
    private var outputReceivers: MutableList<MidiOutputPort> = mutableListOf()
    private var isInitialized = false

    // MIDI Clock
    private val scheduler = Executors.newSingleThreadScheduledExecutor()
    private var clockFuture: ScheduledFuture<*>? = null
    private var clockBpm: Double = 120.0
    @Volatile private var isSendingClock = false

    fun initialize(): Boolean {
        if (isInitialized) return true
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            Log.w(TAG, "MIDI not supported below Android 6.0")
            return false
        }

        midiManager = context.getSystemService(Context.MIDI_SERVICE) as? MidiManager
        if (midiManager == null) {
            Log.e(TAG, "MIDI not available on this device")
            return false
        }

        // Register device callback for hot-plug
        midiManager?.registerDeviceCallback(object : MidiManager.DeviceCallback() {
            override fun onDeviceAdded(device: MidiDeviceInfo) {
                Handler(Looper.getMainLooper()).post {
                    onEvent(mapOf(
                        "type" to "midiDeviceChange",
                        "action" to "added",
                        "devices" to listDevices()
                    ))
                }
            }
            override fun onDeviceRemoved(device: MidiDeviceInfo) {
                Handler(Looper.getMainLooper()).post {
                    onEvent(mapOf(
                        "type" to "midiDeviceChange",
                        "action" to "removed",
                        "devices" to listDevices()
                    ))
                }
            }
        }, Handler(Looper.getMainLooper()))

        isInitialized = true
        Log.d(TAG, "MIDI initialized")
        return true
    }

    fun listDevices(): List<Map<String, Any>> {
        val manager = midiManager ?: return emptyList()
        return manager.devices.map { info ->
            val props = info.properties
            mapOf(
                "name" to (props.getString(MidiDeviceInfo.PROPERTY_NAME) ?: "Unknown"),
                "manufacturer" to (props.getString(MidiDeviceInfo.PROPERTY_MANUFACTURER) ?: ""),
                "id" to info.id,
                "inputCount" to info.inputPortCount,
                "outputCount" to info.outputPortCount,
                "type" to when (info.type) {
                    MidiDeviceInfo.TYPE_USB -> "usb"
                    MidiDeviceInfo.TYPE_BLUETOOTH -> "bluetooth"
                    MidiDeviceInfo.TYPE_VIRTUAL -> "virtual"
                    else -> "unknown"
                }
            )
        }
    }

    fun connectToDevice(deviceId: Int) {
        val manager = midiManager ?: return
        val deviceInfo = manager.devices.find { it.id == deviceId } ?: return

        manager.openDevice(deviceInfo, { device ->
            if (device == null) {
                Log.e(TAG, "Failed to open MIDI device $deviceId")
                return@openDevice
            }
            openDevices.add(device)

            // Open input ports (for sending MIDI TO the device)
            for (i in 0 until device.info.inputPortCount) {
                val port = device.openInputPort(i)
                if (port != null) inputPorts.add(port)
            }

            // Open output ports (for receiving MIDI FROM the device)
            for (i in 0 until device.info.outputPortCount) {
                val port = device.openOutputPort(i)
                if (port != null) {
                    outputReceivers.add(port)
                    port.connect(object : MidiReceiver() {
                        override fun onSend(data: ByteArray, offset: Int, count: Int, timestamp: Long) {
                            processMidiData(data, offset, count)
                        }
                    })
                }
            }

            Log.d(TAG, "Connected to MIDI device: ${deviceInfo.properties.getString(MidiDeviceInfo.PROPERTY_NAME)}")
        }, Handler(Looper.getMainLooper()))
    }

    // MARK: - Send MIDI

    fun sendNoteOn(note: Int, velocity: Int, channel: Int = 0) {
        sendBytes(byteArrayOf(
            (0x90 or (channel and 0x0F)).toByte(),
            (note and 0x7F).toByte(),
            (velocity and 0x7F).toByte()
        ))
    }

    fun sendNoteOff(note: Int, channel: Int = 0) {
        sendBytes(byteArrayOf(
            (0x80 or (channel and 0x0F)).toByte(),
            (note and 0x7F).toByte(),
            0
        ))
    }

    fun sendCC(controller: Int, value: Int, channel: Int = 0) {
        sendBytes(byteArrayOf(
            (0xB0 or (channel and 0x0F)).toByte(),
            (controller and 0x7F).toByte(),
            (value and 0x7F).toByte()
        ))
    }

    fun sendProgramChange(program: Int, channel: Int = 0) {
        sendBytes(byteArrayOf(
            (0xC0 or (channel and 0x0F)).toByte(),
            (program and 0x7F).toByte()
        ))
    }

    private fun sendBytes(data: ByteArray) {
        for (port in inputPorts) {
            try {
                port.send(data, 0, data.size)
            } catch (e: Exception) {
                Log.e(TAG, "Send MIDI error: $e")
            }
        }
    }

    // MARK: - MIDI Clock

    fun startClock(bpm: Double) {
        stopClock()
        clockBpm = bpm
        isSendingClock = true

        // Send MIDI Start
        sendBytes(byteArrayOf(0xFA.toByte()))

        // 24 ppq clock
        val intervalUs = ((60_000_000.0 / bpm / 24.0)).toLong()
        clockFuture = scheduler.scheduleAtFixedRate({
            if (isSendingClock) {
                sendBytes(byteArrayOf(0xF8.toByte()))
            }
        }, 0, intervalUs, TimeUnit.MICROSECONDS)
    }

    fun stopClock() {
        isSendingClock = false
        clockFuture?.cancel(false)
        clockFuture = null
        sendBytes(byteArrayOf(0xFC.toByte())) // MIDI Stop
    }

    fun updateClockBpm(bpm: Double) {
        if (isSendingClock) {
            startClock(bpm)
        } else {
            clockBpm = bpm
        }
    }

    // MARK: - Receive Processing

    private fun processMidiData(data: ByteArray, offset: Int, count: Int) {
        var i = offset
        while (i < offset + count) {
            val status = data[i].toInt() and 0xFF
            val msgType = status and 0xF0
            val channel = status and 0x0F

            val event = mutableMapOf<String, Any>(
                "type" to "midi",
                "status" to status,
                "channel" to channel
            )

            when {
                msgType == 0x90 && i + 2 < offset + count -> {
                    val note = data[i + 1].toInt() and 0x7F
                    val vel = data[i + 2].toInt() and 0x7F
                    event["kind"] = if (vel > 0) "noteOn" else "noteOff"
                    event["note"] = note
                    event["velocity"] = vel
                    i += 3
                }
                msgType == 0x80 && i + 2 < offset + count -> {
                    event["kind"] = "noteOff"
                    event["note"] = data[i + 1].toInt() and 0x7F
                    event["velocity"] = data[i + 2].toInt() and 0x7F
                    i += 3
                }
                msgType == 0xB0 && i + 2 < offset + count -> {
                    event["kind"] = "cc"
                    event["controller"] = data[i + 1].toInt() and 0x7F
                    event["value"] = data[i + 2].toInt() and 0x7F
                    i += 3
                }
                msgType == 0xC0 && i + 1 < offset + count -> {
                    event["kind"] = "programChange"
                    event["program"] = data[i + 1].toInt() and 0x7F
                    i += 2
                }
                status == 0xF8 -> { event["kind"] = "clock"; i += 1 }
                status == 0xFA -> { event["kind"] = "start"; i += 1 }
                status == 0xFB -> { event["kind"] = "continue"; i += 1 }
                status == 0xFC -> { event["kind"] = "stop"; i += 1 }
                else -> { i += 1 }
            }

            Handler(Looper.getMainLooper()).post { onEvent(event) }
        }
    }

    // MARK: - Cleanup

    fun dispose() {
        stopClock()
        for (port in inputPorts) { try { port.close() } catch (_: Exception) {} }
        for (port in outputReceivers) { try { port.close() } catch (_: Exception) {} }
        for (device in openDevices) { try { device.close() } catch (_: Exception) {} }
        inputPorts.clear()
        outputReceivers.clear()
        openDevices.clear()
        isInitialized = false
        Log.d(TAG, "MIDI disposed")
    }
}
